url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/org/jgraph/graph/VertexRenderer.java

```diff
@@ -0,0 +1,374 @@
+			// JDK Bug: Zero length string passed to TextLayout constructor
+		}
+	}
+
+	/**
+	 * Provided for subclassers to paint a selection border.
+	 */
+	protected void paintSelectionBorder(Graphics g) {
+		Graphics2D g2 = (Graphics2D) g;
+		Stroke previousStroke = g2.getStroke();
+		g2.setStroke(GraphConstants.SELECTION_STROKE);
+		if (childrenSelected || selected) {
+			if (childrenSelected)
+				g.setColor(gridColor);
+			else if (hasFocus && selected)
+				g.setColor(lockedHandleColor);
+			else if (selected)
+				g.setColor(highlightColor);
+			Dimension d = getSize();
+			g.drawRect(0, 0, d.width - 1, d.height - 1);
+		}
+		g2.setStroke(previousStroke);
+	}
+
+	/**
+	 * Returns the intersection of the bounding rectangle and the straight line
+	 * between the source and the specified point p. The specified point is
+	 * expected not to intersect the bounds.
+	 */
+	public Point2D getPerimeterPoint(VertexView view, Point2D source, Point2D p) {
+		Rectangle2D bounds = view.getBounds();
+		double x = bounds.getX();
+		double y = bounds.getY();
+		double width = bounds.getWidth();
+		double height = bounds.getHeight();
+		double xCenter = x + width / 2;
+		double yCenter = y + height / 2;
+		double dx = p.getX() - xCenter; // Compute Angle
+		double dy = p.getY() - yCenter;
+		double alpha = Math.atan2(dy, dx);
+		double xout = 0, yout = 0;
+		double pi = Math.PI;
+		double pi2 = Math.PI / 2.0;
+		double beta = pi2 - alpha;
+		double t = Math.atan2(height, width);
+		if (alpha < -pi + t || alpha > pi - t) { // Left edge
+			xout = x;
+			yout = yCenter - width * Math.tan(alpha) / 2;
+		} else if (alpha < -t) { // Top Edge
+			yout = y;
+			xout = xCenter - height * Math.tan(beta) / 2;
+		} else if (alpha < t) { // Right Edge
+			xout = x + width;
+			yout = yCenter + width * Math.tan(alpha) / 2;
+		} else { // Bottom Edge
+			yout = y + height;
+			xout = xCenter + height * Math.tan(beta) / 2;
+		}
+		return new Point2D.Double(xout, yout);
+	}
+
+	/**
+	 * Overridden for performance reasons. See the <a
+	 * href="#override">Implementation Note </a> for more information.
+	 */
+	public void validate() {
+	}
+
+	/**
+	 * Overridden for performance reasons. See the <a
+	 * href="#override">Implementation Note </a> for more information.
+	 */
+	public void revalidate() {
+	}
+
+	/**
+	 * Overridden for performance reasons. See the <a
+	 * href="#override">Implementation Note </a> for more information.
+	 */
+	public void repaint(long tm, int x, int y, int width, int height) {
+	}
+
+	/**
+	 * Overridden for performance reasons. See the <a
+	 * href="#override">Implementation Note </a> for more information.
+	 */
+	public void repaint(Rectangle r) {
+	}
+
+	/**
+	 * Overridden for performance reasons. See the <a
+	 * href="#override">Implementation Note </a> for more information.
+	 */
+	protected void firePropertyChange(String propertyName, Object oldValue,
+			Object newValue) {
+		// Strings get interned...
+		if (propertyName == "text")
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.